### ISSUE TITLE
Episodes in grid

### ISSUE THREAD

**Issue by rmcrackan** [rmcrackan/Libation#154]
## New way to display episodes with parents

Currently every episode is shown individually and is a complete eyesore. Episodes without parent: still show as a normal row. Episodes with parent:

Parent has single grid entry/row. Replace stoplight with "+" or "-" image. Button expands/collapses children in new rows beneath. Actually just shows or hides them. They'll need to already be in the correct order (addressed below).

Only parents are affected by sorting and filtering.
* Sort: When expanded, episodes always appear in series order below parent irrespective of other grid sorting
* Filter: If parent is shown, all children are available to be shown. If parent is hidden, all children are hidden

sorting idea 1: https://www.codeproject.com/Articles/18440/DataGridView-Multi-column-Sort

sorting idea 2: bindingSource.Sort = "[Street Name], [Street Direction], [Street Number]";

### Sorting example
* Metamorphosis. podcast analyzing Kafka. child. episode 1. series id 789
* Metamorphosis. podcast butterflies. child. episode 2. series id 123
* ABC book
* Metamorphosis. podcast butterflies. parent. series id 123
* XYZ book
* Metamorphosis. podcast butterflies. child. episode 1. series id 123
* Metamorphosis. book by Ovid. no episode series
* Metamorphosis. podcast analyzing Kafka. child. episode 2. series id 789
* Metamorphosis. podcast analyzing Kafka. parent. series id 789

#### Sort level 1: user-defined. Eg: click column

CreateMemberValueDictionary() will have to use same values for parents and all children

* **ABC book**
* Metamorphosis. podcast analyzing Kafka. child. episode 1. series id 789
* Metamorphosis. podcast butterflies. child. episode 2. series id 123
* Metamorphosis. podcast butterflies. parent. series id 123
* Metamorphosis. podcast butterflies. child. episode 1. series id 123
* Metamorphosis. book by Ovid. no episode series
* Metamorphosis. podcast analyzing Kafka. child. episode 2. series id 789
* Metamorphosis. podcast analyzing Kafka. parent. series id 789
* **XYZ book**

#### Sort level 2: episode series. User has no control over this and subsequent sort levels

* ABC book
* **Metamorphosis. book by Ovid. no episode series**
* Metamorphosis. podcast butterflies. child. episode 2. series id 123
* Metamorphosis. podcast butterflies. parent. series id 123
* Metamorphosis. podcast butterflies. child. episode 1. series id 123
* Metamorphosis. podcast analyzing Kafka. child. episode 1. series id 789
* Metamorphosis. podcast analyzing Kafka. child. episode 2. series id 789
* Metamorphosis. podcast analyzing Kafka. parent. series id 789
* XYZ book

#### Sort level 3: is parent

* ABC book
* Metamorphosis. book by Ovid. no episode series
* **Metamorphosis. podcast butterflies. parent. series id 123**
* Metamorphosis. podcast butterflies. child. episode 2. series id 123
* Metamorphosis. podcast butterflies. child. episode 1. series id 123
* **Metamorphosis. podcast analyzing Kafka. parent. series id 789**
* Metamorphosis. podcast analyzing Kafka. child. episode 1. series id 789
* Metamorphosis. podcast analyzing Kafka. child. episode 2. series id 789
* XYZ book

#### Sort level 4: series order ascending

* ABC book
* Metamorphosis. book by Ovid. no episode series
* Metamorphosis. podcast butterflies. parent. series id 123
* **Metamorphosis. podcast butterflies. child. episode 1. series id 123**
* **Metamorphosis. podcast butterflies. child. episode 2. series id 123**
* Metamorphosis. podcast analyzing Kafka. parent. series id 789
* **Metamorphosis. podcast analyzing Kafka. child. episode 1. series id 789**
* **Metamorphosis. podcast analyzing Kafka. child. episode 2. series id 789**
* XYZ book

Outstanding concerns:

* How will parent vs child be indexed by search engine?
* How will parent vs child be stored in db? Is this a strong enough argument to retain rdbs?

**Comment by Mbucari** [comment]
So it looks like there is a framework for doing this already built into .Net. https://www.c-sharpcorner.com/UploadFile/deveshomar/expandable-and-collapsible-rows-in-datagrid-in-C-Sharp-winforms/

I'm not sure how this is supposed to be implemented with MVVM, but I'm certain it will involve changes to GridView. I think this is what I'm going to try to tackle next.

**Comment by rmcrackan** [comment]
That component is for DataGrid instead of DataGridView. I'm not married to DGV but it's a much more versatile swiss army knife. If all current features are compatible with that new guy, then go for it. But I imagine that shoehorning some of the other stuff we've done will be more work than a new implementation for DGV.

Everything in my original post is a scratchpad of ideas and proposed implementation details -- nothing written in stone. The essence of this feature request is the top part.

* Single row parent/series with a way to collapse and expand their children episodes. Since parents aren't real items, we can commandeer their stoplight area for the +/- stuff
* Current sorting and filtering needs to still work. Parents are sorted and filtered per that paradigm. Children should not to be sorted or filtered individually. They should do so as defined above

**Comment by Mbucari** [comment]
Yup, I didn't notice that last night.  I shouldn't try planning while drinking XD

This project seems to fit the bill quite nicely though.
https://www.codeproject.com/Tips/1215736/Master-Detail-Datagridview-in-Csharp

**Comment by rmcrackan** [comment]
> I shouldn't try planning while drinking

We're not going to talk about how much of Libation I wrote while sipping bourbon, beer, or red wine. The name Libation really is just a play on LIBerate, LIBrary, LIBation -- buuuuut it's also something with which I have some experience

That codeproject article looks nice. It could save us some time. If it doesn't work out, it might not be too difficult to implement by hand (says the guy who hasn't written it yet). We know how to change images, insert at index, remove at index, hide, sort, and filter. We also already have code wired up to all of those functions, so we know where the entry points are.

**Comment by Mbucari** [comment]
I've already downloaded that project and played around with it a bit, and I'm pretty convinced we need to build our own. It is helpful as a proof of concept though.

The way I see this working as adding a List<GridEntry> member to GridEntry. For a GridEntry that is a parent series containing episodes, populate that list with all of its episodes. And inside LiberateDataGridViewImageButtonColumn, show a plus sign for all GridEntry rows that have items in that list. Clicking that button then expands the row to show a second DataGridView of all the children.

That approach solves the sorting problem inherently, because the master DataGridView's data source will not contain any episodes. Episodes are only added to the details DataGridView on demand when it is expanded by clicking on the plus sign.

But filtering Will still be a problem because, presumably, the user would like the ability to search individual episodes. But if the master data source doesn't contain any episodes, something will have to be jury rigged to include those items within the search.

**Comment by rmcrackan** [comment]
> That approach solves the sorting problem inherently

Good. My idea is solid but relies on algorithm rather than encapsulation. I much prefer the cleanness of separation/encapsulation.

> presumably, the user would like the ability to search individual episodes

I've been thinking about this all weekend and I can come up with an argument either way. I *think* the more common case is just parent/series. Either way we're going to have to change things and we can only pick one. The hard part for all this is that search/filter was designed with the paradigm that all titles are equal, but now we're introducing hierarchy. I believe the easiest solution would be to make series all or nothing:

I can remove episodes from the search engine; add parents to it. Search applies only to parents. When a search includes a parent, fetch all of its children. Search flags like `ispodcast` will still apply to the parent. This also supports the db ideas that episodes are books and parents are series. IsDownloaded will still apply to children. Children can be tagged (series cannot be). Tags will actually be rolled up into their parent's search engine entry.

btw -- the new grid binding seems to have a bug when new items come in. Fortunately it's a refresh bug so it's fixed by changing the filter to something else then changing it back: filter `-israted` ie: omit books with a user rating. Scan. Existing book correctly refreshes user rating. Bug: but the row should now be hidden because of the filter.

**Comment by Mbucari** [comment]
> I much prefer the cleanness of separation/encapsulation.

Unfortunately that project just wouldn't work.  The problem is that DataGridView can't display a partial row.  The row is either all shown or all hidden.  So a second DataGridView inside an expanded row can not be able to be very large, and it becomes unusable if it's inside the last row of the master DataGridView.  Plus, that solution required a LOT of Paint() hacking.  The algorithmic approach, for as much of a pain in the ass it was to write and will be to debug, is actually cleaner.

> I believe the easiest solution would be to make series all or nothing

That would have been easier, bu it's not what I did.  I implemented partial series. There are a lot of edge cases here, and I think it's going to take a lot of testing to find the all.  I'll need your help with that.

> btw -- the new grid binding seems to have a bug when new items come in.
Fixed in #255
